docs/semgrep-code/remove-duplicates.md

```diff
@@ -3,29 +3,111 @@ slug: remove-duplicates
+ <dt>Full scan</dt>
+ <dd>Scans the repository in its entirety. It is recommended to perform full scans on mainline branches, such as <code>master</code> or <code>main</code>. Full scans are typically performed on a scheduled basis or on merge to a default branch.</dd>
+ <dt>Diff-aware scan</dt>
+ <dd>Diff-aware scans are performed on non-mainline branches, such as in pull requests and merge requests. Diff-aware scans traverse the repository's files based on the commit where the branch diverged from the mainline branch.</dd>
 </dl>
 
-## Remove duplicate findings using Semgrep AppSec Platform
+## How Semgrep distinguishes between new and duplicate findings
 
-Regardless of the scope of a scan, Semgrep correlates findings across branches based on their unique fingerprint, automatically deduplicating findings and making it simpler to triage.
+Semgrep generates a finding whenever it scans a repository, and one of its rules matches a piece of code. Since Semgrep scans a repository multiple times, it needs a way to track the same finding in a file over time. Semgrep does this using two types of fingerprints: `match_based_id` and `syntactic_id`.
 
-If a finding is fixed in one branch (such as `main`) but open in another (such as `production`), and the code fixes are present in both branches, initiate a scan through your CI job or SCM tool on the branch(es) with open findings to have Semgrep mark the findings as fixed.
+:::info
+The calculations used to determine whether findings are new are subject to change at any time as Semgrep improves its deduplication logic.
+:::
+
+### `match_based_id`
+
+Semgrep generates the `match_based_id`, the primary fingerprint used to track and deduplicate findings, for a finding using the following information:
+
+- The file path
+- The name of the rule that generated the finding
+- The rule pattern with the metavariables' values substituted in
+
+This information is combined and then hashed. At this point, Semgrep appends the **index**, a value generated by determining the number of times the rule involved matched code in the file. Note that the index is appended to the hash, not combined with the other finding information before hashing. This is done to preserve information on how findings are related. For example, `finding0` with `match_based_id = 123_0` and `finding1` with `match_based_id = 123_1` indicate that both were generated from the same rule matching the same code pattern in the same file.
+
+Because Semgrep uses the rule pattern instead of the literal code syntax to generate `match_based_id`, code changes that don't impact the code pattern matched also don't hinder Semgrep's ability to recognize that the finding isn't a duplicate of an existing finding.
```

Comment:
This is definitely much closer, but our usage of "rule pattern" and "code pattern" here is a bit inconsistent. The content of the code could still be different here, but when abstracted to the level of the match with the substituted values then that's what's the same. But this ischallenging because "rule pattern with the metavariables' values [from the code] substituted in" (which is the actual thing that's the same) is cumbersome to say.
@pangngan do you have ideas about how best to express this?

Reply:
@armchairlinguist I don't know if this is any better, but

> Semgrep uses the rule pattern with the metavariables' values, which originate from the code itself, substituted in to generate `match_based_id`. Therefore, code changes that result in the same rule pattern when abstracted to the level of the rule pattern match, with the appropriate values substituted, in don't hinder Semgrep's ability to recognize that the finding isn't a duplicate of an existing finding.